# Overview

This release brings full support of sample barcodes in MiXCR pipeline. Sample barcodes can be picked up and processed
from all possible sources (`I1`/`I2` files, sequence header lines, from inside the tag pattern, etc...). Processing can
be done in two principal modes in respect to sample barcodes: (1) data can be split by samples right on the `align`
stage and processed separately, or (2) all samples can be processed as a single set of sequences and separated only on
the very last `exportClones` step, both approaches have their pros and cons allowing to use the best strategy given the
experimental setup and study goals.

# Major changes

- support for more than two `fastq` files as input (`I1` and `I2` reads support)
- multiple possible sources of data for sample resolution:
  - sequences extracted with tag pattern (including those coming from `I1` and `I2` reads)
  - samples can be based on specific pattern variant (with multi-variant patterns, separated by `||`, allows to easily
    adopt MiGEC-style sample files)
  - [TBD] strings extracted from fastq description line
  - parts of file names (extracted using file name expansion mechanism)
- flexible sample table matching criteria
  - matching multiple tags
  - matching variant id from multi-variant tag patterns
- `align` command now optionally allows to split output alignments by sample into separate `vdjca` files
- `exportClones` command now supports splitting the output into multiple files by sample
- `analyze` command supports new splitting behaviour of the `align` command, separately running all the analysis steps
  for all the output files (if splitting is enabled)
- added command `exportReportsTable` that prints file in tabular format with report data from commands that were run.

# Minor changes & fixes

- add tags info in `exportAlignmentsPretty` and `exportClonesPretty`
- add `--chains` filter for `exportShmTrees`, `exportShmTreesWithNodes`, `exportShmTreesNewick`
  and `exportPlots shmTrees` commands
- fixed old bug #353, now all aligners favor leftmost J gene in situations where multiple genes can ve found in the 
  sequence (i.e. mis-spliced mRNA)
- new filter and parameter added in `assemblePartial`; parameter name is `minimalNOverlapShare`, it controls minimal
  relative part of N region that must be covered by the overlap to conclude that two reads are from the same V(D)J
  rearrangement
- fixed NPE in `assemblePartial` executed for the data without C-gene alignment settings
- by default exports show messages like 'region_not_covered' for data that can't be extracted (requesting `-nFeature`
  for not covered region or not existed tag). Option `--not-covered-as-empty` will save previous behaviour
- info about genes with enough data to find allele was added into report of `findAlleles` and description of alleles
