- new presets for NEBNextÂ® Immune Sequencing Kit TCR and BCR profiling for data with both TCR and BCR.
- improved Takara human BCR presets
- added `discardAmbiguousNucleotideCalls` parameters for contig assembly
- new group filter operators allowing to mix thresholds form multiple operators, taking lowest or highest value and
  applying it. This allows to create more universal filtering strategies, robust to edge cases like undersequencing of
  barcodes.
- threshold rounding in cumtop and top-n filters
- Improved Otsu's method with less stringency for automated histogram thresholding for barcoded data. It allows to
  recover more "good" UMI groups. The old filter was replaced by new one in all presets for airr-seq and single cell vdj
  protocols that utilize UMI: Cellecta, Milaboratories, NEB, Qiagen, Takara, 10x Genomics, BD, Singleron.   
- better text descriptions in align and assemble reports

## Minor improvements

- added field `-cellId` in commands `exportClones` and `exportAlignments`
- added fields `cell_id`, `umi_count` and `consensus_count` to `exportAirr` command
- `exportAirr` command now split clones by cells if there is cell barcodes in the data
- in `analyze` options `--not-aligned-..` and `--not-parsed-..` replaced with one option `--output-not-used-reads`
- fix comma-separated chains input in postanalysis --chains option
- `#` character now can be used to separate groupName from group matcher in file expansion mechanist (additionally to `:`)
