## Key changes

- implemented QC [TODO]
- new alignment
- special mechanism to allow for NaN values in metrics in group filters (i.e. used in `minSequenceCount` parameter in k-mer filter, see below)
- adds new `minSequenceCount` parameter for k-mer filter, allowing construction of more flexible filtering pipelines with better fallback behaviour for under-sequenced libraries
- now full sample sheet with input file names can be provided as an input to the pipeline
- now sample sheets provided both with `--sample-table` mixin and as a pipeline input, will be fuzzy matched against the data, allowing for one substitutions in unambiguous cases. This behaviour can be turned off by using `--sample-table-strict` mixin instead, or by adding a `--strict-sample-sheet-matching` option if full sample sheet input is used as pipeline input. 

## Major fixes

- fix for naming of intermediate files and reports produced by `analyze` if target folder is specified
- tag pattern now is also searched in reverse strand for single-ended input with `--tag-parse-unstranded`

## Preset changes

- added fallback behaviour for under-sequenced libraries for `10x-vdj-bcr` preset

## Minor fixes and improvements

- added gene feature coverage in alignment report
- new gene aligner parameter `edgeRealignmentMinScoreOverride` for more sensitive alignments for short paired-end reads 
- report values downstream `align` now calculate percents relative to the number of reads in the sample rather than the total number of reads in multi-sample analysis
- fixed bug not allowing to parse more than two reads with tag pattern
- better tag pattern search projection representation in reports
- fix for value in report line `Reads dropped due to low quality, percent of total report string`
- additional report string "Aligned reads processed" in `assemble` report
- fix bug when `--chains` is used with `exportClonesOverlap` 
- fix for `export...` - tag quality field added back to export columns
- fixes bug not allowing to use more than 2 input reads in pattern matching
