## ‚ùó Breaking changes

- Starting from version 4.7.0 of MiXCR, users are required to specify the assembling feature for all presets in cases
  where it's not defined by the protocol. This can be achieved using either the option `--assemble-clones-by Feature`
  or `--assemble-contigs-by Feature` for fragmented data (such as RNA-seq or 10x VDJ data). This ensures consistency in
  assembling features when integrating various samples or types of samples, such as 10x single-cell VDJ and AIRR
  sequencing data, for downstream analyses like inferring alleles or building SHM trees. The previous behavior for
  fragmented data, which aimed to assemble as long sequences as possible, can still be achieved with either the
  option `--assemble-contigs-by-cell` for single-cell data or `--assemble-longest-contigs` for RNA-seq/Exom-seq data.

## üöÄ Major fixes and upgrades

- Ability to trigger realignments of left or right reads boundaries with global alignment algorythm using
  parameters `rightForceRealignmentTrigger Feature` or `leftForceRealignmentTrigger Feature` in case the reads do
  not span the CDR3 regions (rescue alignments in case of fragmented single cell data).
- Fixed `assemble` behavior in presets for single-cell data (in some cases consensuses were assembled from reads coming
  from different cells)
- Ability to override the `relativeMeanScore` and `maxHits` parameters in `assemble` and `assembleContigs` steps
  (improve the V genes assignments)
- Consensus assembly in `assemble` now is performed separately for each chain. This allows to prevent effects from
  different expression levels on the consensus assembly algorithm. This change is specifically important for single-cell
  presets with cell-level assembly (most of the MiXCR presets for single-cell data).
- Options `--dont-correct-tag-with-name <tag_name>` or `--dont-correct-tag-type (Molecule|Cell|Sample)` could be
  specified to skip tag correction. It will degrade the overall quality of analysis, but will decrease memory
  consumption
- MiTool pipeline integrated into `10x-sc-xcr-vdj` preset which improved overall quality of `analyze`

## üõ†Ô∏è Minor improvements & fixes

- Default input quality filter in `assemble`  (`badQualityThreshold`) stage was decreased to 10.
- Added validation for `assembleCells` that input files should be assembled by fixed feature
- Export of trees and tree nodes now support imputed features
- Fixed parsing of optional arguments
  for `exportShmTreesWithNodes`: `-nMutationsRelative`, `-aaMutations`, `-nMutations`, `-aaMutationsRelative`, `-allNMutations`, `-allAAMutations`, `-allNMutationsCount`, `-allAAMutationsCount`.
- Fixed parsing of optional arguments for `exportClones`
  and `exportAlignments`: `-allNMutations`, `-allAAMutations`, `-allNMutationsCount`, `-allAAMutationsCount`.
- Fixed possible errors on exporting amino acid mutations in `exportShmTreesWithNodes`
- Fixed list of required options in `listPresets` command
- Fixed error on building trees in case of JBeginTrimmed started before CDR3Begin
- Fixed usage `--remove-step qc`
- Added `--remove-qc-check` option
- Remove `-topChains` field from `exportShmTreesWithNodes` command. Use `-chains` instead
- Removed default splitting clones by V and J for presets where clones are assembled by full-length.
- Fixed NullPointerException in some cases of building trees by SC+bulk data
- Fixed `java.lang.IllegalArgumentException: While adding VEndTrimmed` in `exportClones`
- Fixed combination trees step in `findShmTrees`: in some cases trees weren't combined even if it could be
- Fixed `java.util.NoSuchElementException` in some cases of SC combining of trees
- Fixed export of `-jBestIdentityPercent` in `exportShmTreesWithNodes`
- Added validation on export `-aaFeature` for features containing UTR
- Fixed usage of command `exportPlots shmTrees`
- Fixed topology of trees: before common V and J mutations were included in the root node, now in root included only
  reconstructed NDN. Previous behavior lead to smaller numbers of distance from germline, sequence for germline exported
  with common mutations. For fix this you should recalculate `findShmTrees`
- Fixed `IllegalStateException` on removing unnecessary genes on `findAlleles`
- Added `--dont-remove-unused-genes` option to `findAlleles` command
- Adjustment consensus assembly (in `assemble`) parameters for single cell presets
- Command `groupClones` was renamed to `assembleCells`. Old name is working, but it's hidden from help. Also report and
  output file names in `analyze` step were renamed accordingly.
- Fixed calculation of germline for `VCDR3Part` and `JCDR3Part` in case of indels inside CDR3
- Fixed export of trees if data assembled by a feature with reference point having offset
- Export of `VJJunction gemline` in `shmTrees` exports now export `mrca` as most plausible content
- Fixed parsing and alignment of reads with length more than 30Kbase
- `downsample` now supports `molecule` variant in `--downsampling` option
- Fixed naming of output files of `downsample` command
- `--output-not-used-reads` of `analyze` command now works with bam input files too, alongside `--not-aligned-(R1|R2)`
  and `--not-parsed-(R1|R2)` of `align` command
- Fix `replaceWildcards` behaviour on parsing BAM. It led before to discarding of quality on `align`
- `v_call`, `d_call`, `j_call` and `c_call` columns in airr now output only bets hit, not the whole list
- Stable behavior of `replaceWildcards`. Before it depended on the position of read in a file, now it depends on read
  content
